File: java/amp-eventing/src/main/java/amp/eventing/RpcRegistration.java
package amp.eventing;


import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeoutException;

import cmf.bus.Envelope;
import cmf.bus.IEnvelopeFilterPredicate;
import cmf.bus.IRegistration;
import org.apache.commons.lang.time.StopWatch;
import org.joda.time.Duration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import amp.eventing.EnvelopeHelper;


public class RpcRegistration implements IRegistration {

    protected static final Logger log = LoggerFactory.getLogger(RpcRegistration.class);

    
    protected IInboundProcessorCallback envelopeOpener;
    protected Map<String, String> registrationInfo;
    protected Envelope responseEnvelope;
    protected Object responseEvent;
    protected IEnvelopeFilterPredicate responseFilter;

    
    public RpcRegistration(final UUID requestId, String expectedTopic, IInboundProcessorCallback envelopeOpener) {
    	
        this.envelopeOpener = envelopeOpener;

        responseFilter = new IEnvelopeFilterPredicate() {

            @Override
            public boolean filter(Envelope envelope) {
                return new EnvelopeHelper(envelope).getCorrelationId().equals(requestId);
            }
        };

        registrationInfo = new HashMap<String, String>();
        Envelope tmpInfoEnvelope = new Envelope();
        tmpInfoEnvelope.setHeaders(registrationInfo);
        new EnvelopeHelper(tmpInfoEnvelope).setMessageTopic(buildRpcTopic(expectedTopic, requestId));
    }

    
    @Override
    public IEnvelopeFilterPredicate getFilterPredicate() {
        return responseFilter;
    }

    @Override
    public Map<String, String> getRegistrationInfo() {
        return registrationInfo;
    }

    public Object getResponse(Duration timeout) throws TimeoutException {
        StopWatch watch = new StopWatch();
        watch.start();
        while (watch.getTime() < timeout.getMillis() && responseEvent == null) {
            try {
                Thread.sleep(Math.min(50, timeout.getMillis() / 10));
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }
        watch.stop();

        Object response = responseEvent;
        responseEvent = null;

        return response;
    }

    @Override
    public Object handle(Envelope envelope) throws Exception {
        
    	log.debug("enter handle");
    	
    	this.responseEnvelope = envelope;
    	this.responseEvent = this.envelopeOpener.ProcessInbound(envelope);
    	log.info("inbound envelope successfully opened");
    	
    	// we have no custom response to give
    	return null;
    }

    @Override
    public Object handleFailed(Envelope envelope, Exception ex) throws Exception {
        log.error("Failed to handle an envelope: " + new EnvelopeHelper(envelope).flatten(), ex);

        return null;
    }

    
    protected void setRegistrationInfo(Map<String, String> registrationInfo) {
        this.registrationInfo = registrationInfo;
    }

    protected String buildRpcTopic(String expectedTopic, UUID requestId) {
        return String.format("%s#%s", expectedTopic, requestId.toString());
    }
}

File: dotnet/source/amp.eventing/RpcRegistration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

using Common.Logging;

using cmf.bus;
using amp.bus;

namespace amp.eventing
{
    public class RpcRegistration : IRegistration
    {
        protected Predicate<Envelope> _responseFilter;
        protected Func<Envelope, object> _openEnvelope;
        protected AutoResetEvent _responseEvent;
        protected ILog _log;
        protected Envelope _responseEnvelope;


        public Predicate<Envelope> Filter
        {
            get { return _responseFilter; }
        }

        public IDictionary<string, string> Info { get; protected set; }
        
        
        public RpcRegistration(Guid requestId, string expectedTopic, Func<Envelope, object> openEnvelope)
        {
            _openEnvelope = openEnvelope;

            _responseEvent = new AutoResetEvent(false);
            _responseFilter = env => Guid.Equals(env.GetCorrelationId(), requestId);
            
            this.Info = new Dictionary<string, string>();
            this.Info.SetMessageTopic(this.BuildRpcTopic(expectedTopic, requestId));

            _log = LogManager.GetLogger(this.GetType());
        }


        public virtual object Handle(Envelope env)
        {
            _responseEnvelope = env;
            _responseEvent.Set();

            return null;
        }

        public virtual object HandleFailed(Envelope env, Exception ex)
        {
            _log.Error("Failed to handle an envelope: " + env.Headers.Flatten(), ex);

            return null;
        }

        public virtual object GetResponse(TimeSpan timeout)
        {
            object response = null;

            if (_responseEvent.WaitOne(timeout))
            {
                response = _openEnvelope(_responseEnvelope);
            }

            return response;
        }


        protected virtual string BuildRpcTopic(string expectedTopic, Guid requestId)
        {
            return string.Format("{0}#{1}", expectedTopic, requestId.ToString());
        }
    }
}
